src/core/bloom.cc

```diff
@@ -104,4 +131,62 @@ inline bool Bloom::Set(size_t bit_idx) {
   return bf_[byte_idx] != b;
 }
 
+///////////////////////////////////////////////////////////////////////////////
+// SBF implementation
+///////////////////////////////////////////////////////////////////////////////
+SBF::SBF(uint64_t initial_capacity, double fp_prob, double grow_factor, PMR_NS::memory_resource* mr)
+    : filters_(1, mr), grow_factor_(grow_factor), fp_prob_(fp_prob * kSBFErrorFactor) {
+  filters_.front().Init(initial_capacity, fp_prob_, mr);
+  max_capacity_ = filters_.front().Capacity(fp_prob_);
+}
+
+SBF::~SBF() {
+  PMR_NS::memory_resource* mr = filters_.get_allocator().resource();
+  for (auto& f : filters_)
+    f.Destroy(mr);
+}
+
+bool SBF::Add(std::string_view str) {
+  DCHECK_LT(current_size_, max_capacity_);
+  auto it = filters_.rbegin();  // largest filter
+  auto cur = it++;
+
+  XXH128_hash_t hash = Hash(str);
+  uint64_t fp[2] = {hash.low64, hash.high64};
+
+  // Check for all other filters whether the item exists.
+  for (; it != filters_.rend(); ++it) {
+    if (it->Exists(fp))
+      return false;
+  }
+
+  if (!cur->Add(fp))
+    return false;
```

Comment:
this is tricky because your variable have a large span, but are not modified 🙂 

so you first check [-2...0] if it exists, then you add to -1

```c++
// Check -2 ... 0 (reverse head?)
auto exists = [fp](const auto& bloom) { return bloom->Exists(fp); };
if (any_of(++filters_.crbegin(), filters_.crend(), exists))
    return false;

// Add to tail
if (!filters_.back().Add(fp))
  return false;
```

Reply:
I check all filters besides last. Then I add to last. Should I add more comments?

Reply:
I will rewrite